kafka-client/src/main/java/dev/responsive/kafka/internal/clients/OffsetTracker.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright 2024 Responsive Computing, Inc.
+ *
+ * This source code is licensed under the Responsive Business Source License Agreement v1.0
+ * available at:
+ *
+ * https://www.responsive.dev/legal/responsive-bsl-10
+ *
+ * This software requires a valid Commercial License Key for production use. Trial and commercial
+ * licenses can be obtained at https://www.responsive.dev
+ */
+
+package dev.responsive.kafka.internal.clients;
+
+import java.util.BitSet;
+
+/**
+ * This class allows us to efficiently count the number of events
+ * between two offsets that match a certain condition. This is somewhat
+ * memory efficient in that we can track 100K offsets with ~1.5K longs
+ * (64 bits per long), or roughly 12KB.
+ */
+public class OffsetTracker {
+
+  private BitSet offsets = new BitSet();
+  private long baseOffset;
+
+  public OffsetTracker(long baseOffset) {
+    this.baseOffset = baseOffset;
+  }
+
+  public void mark(final long offset) {
+    if (offset < baseOffset) {
+      throw new IllegalArgumentException(
```

Comment:
nit: can we log an error before throwing that includes the value of `offsets` (here and below)